Fix #62: Show test coverage with makefile

kwk opened the pull request:
Testing has become a rather big and interconnected topic and that's why it
has arrived in it's own file.

We have to types of tests available:

  1. unit tests and
  2. integration tests.

 While the unit tests can be executed fairly simply be running `go test`, the
 integration tests have a little bit more setup going on. That's why they are
split up in two tests.

# Usage

If you want to run the unit tests, type

    $ make test-unit

To run the integration tests, type

     $ make test-integration

 To run both tests, type

     $ make test-all

 To show unit test coverage in a terminal, type

     $ make show-coverage-unit

 To show unit test coverage in a browser, type

     $ make show-coverage-unit-html

 If you replace the "unit" with "integration" you get the same for integration
 tests.

 # Artifacts and coverage modes

 We execute tests with coverage output by default. The files live in the
 root folder and look something like `coverage-unit-mode-set.out` or
 `coverage-integration-mode-count.out`.

 Each filename indicates whether it is for unit or integration tests and
 what the coverage mode was used.

 These are possible coverage modes (see https://blog.golang.org/cover):

 	set: did each statement run? (default)
 	count: how many times did each statement run?
 	atomic: like count, but counts precisely in parallel programs

 To choose another coverage mode, simply prefix the invovation of `make`:

     $ COVERAGE_MODE=count make test-unit

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/almighty/almighty-core/64)
<!-- Reviewable:end -->

aslakknutsen commented:
*[.gitignore, line 26 \[r1\]](https://reviewable.io:443/reviews/almighty/almighty-core/64#-KOA2iDuaG2i171tkEUA:-KOA2iDuaG2i171tkEUB:1150490430) ([raw file](https://github.com/almighty/almighty-core/blob/2db23219938c373843c8e778e2458c80b73483ef/.gitignore#L26)):*
> ```
> # Ignore Go coverage files
> **/coverage-unit.out
> **/coverage-integration.out
> ```

Fresh already use tmp/ to store some state. Should we reuse that folder here as well?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/64)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented:
*[.make/test.mk, line 30 \[r1\]](https://reviewable.io:443/reviews/almighty/almighty-core/64#-KOA39QY0GJrkZ8fR3Qy:-KOA39QY0GJrkZ8fR3Qz:296699132) ([raw file](https://github.com/almighty/almighty-core/blob/9044a75dec88ae0ea999171c6206c7183137b5c3/.make/test.mk#L30)):*
> ```Makefile
> # To show unit test coverage in a terminal, type
> #
> #     $ make show-coverage-unit-func
> ```

Why 'func' ? as oppose to nothing, or -console?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/64)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented:
*[.make/test.mk, line 109 \[r1\]](https://reviewable.io:443/reviews/almighty/almighty-core/64#-KOA3apgCJuz-relcdNP:-KOA3aphOcRdGG_z-hDn:146919199) ([raw file](https://github.com/almighty/almighty-core/blob/9044a75dec88ae0ea999171c6206c7183137b5c3/.make/test.mk#L109)):*
> ```Makefile
> #-------------------------------------------------------------------------------
> 
> $(COVERAGE_UNIT_PATH):
> ```

These are fairly internal to the Makefile. Any way to 'exclude' the target from bash autocomplete?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/64)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented:
*[Makefile, line 55 \[r1\]](https://reviewable.io:443/reviews/almighty/almighty-core/64#-KOA40_h9FMhs3BAmeBS:-KOA40_h9FMhs3BAmeBT:-1864240991) ([raw file](https://github.com/almighty/almighty-core/blob/9044a75dec88ae0ea999171c6206c7183137b5c3/Makefile#L55)):*
> ```Makefile
> 
> .PHONY: clean
> clean: clean-artifacts clean-object-files clean-generated clean-vendor clean-glide-cache clean-test-artifacts-unit clean-test-artifacts-integration
> ```

Is it possible to define something like; clean-* ?

Would be nice if we could have dependencies in only one direction. Currently, if you comment out the test.mk the main mk fail due to missing targets.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/64)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented:
*[.make/test.mk, line 30 \[r1\]](https://reviewable.io:443/reviews/almighty/almighty-core/64#-KOA39QY0GJrkZ8fR3Qy:-KOA5-x1gkYePP7IMfDh:1831884166) ([raw file](https://github.com/almighty/almighty-core/blob/9044a75dec88ae0ea999171c6206c7183137b5c3/.make/test.mk#L30)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> Why 'func' ? as oppose to nothing, or -console?
</details>
Based on bash autocomplete it would be nice if the console option had no 'suffix'. you could auto complete with just: make s<tab>u<tab><enter>

Currently, you need:
make s<tab>u<tab>f<tab><enter>

Suggest:

show-coverage-integration
show-coverage-integration-html 
show-coverage-unit       
show-coverage-unit-html 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/64)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented:
*[.make/test.mk, line 30 \[r1\]](https://reviewable.io:443/reviews/almighty/almighty-core/64#-KOA39QY0GJrkZ8fR3Qy:-KOA5lNNhkuuyLx9kRSD:2123258123) ([raw file](https://github.com/almighty/almighty-core/blob/9044a75dec88ae0ea999171c6206c7183137b5c3/.make/test.mk#L30)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> Based on bash autocomplete it would be nice if the console option had no 'suffix'. you could auto complete with just: make s<tab>u<tab><enter>
> 
> Currently, you need:
> make s<tab>u<tab>f<tab><enter>
> 
> Suggest:
> 
> show-coverage-integration
> show-coverage-integration-html 
> show-coverage-unit       
> show-coverage-unit-html 
> 
</details>
Lost some 'tags' in there.. 

You could auto complete with just: make s'TAB'u'TAB''ENTER'

Currently, you need:
make s'TAB'u'TAB'f'TAB''ENTER'

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/64)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented:
*[.make/test.mk, line 99 \[r1\]](https://reviewable.io:443/reviews/almighty/almighty-core/64#-KOA6xAMaX6haTDmuNG6:-KOA6xANr_Ru1XD9xK8d:-294607742) ([raw file](https://github.com/almighty/almighty-core/blob/9044a75dec88ae0ea999171c6206c7183137b5c3/.make/test.mk#L99)):*
> ```Makefile
> .PHONY: show-coverage-integration-func
> show-coverage-integration-func: prebuild-check $(COVERAGE_INTEGRATION_PATH)
> 	go tool cover -func=$(COVERAGE_UNIT_PATH)
> ```

That should be func=${COVERAGE_INTEGRATION_PATH}

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/64)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented:
*[a discussion](https://reviewable.io:443/reviews/almighty/almighty-core/64#-KOAA7VJ0kFmgj8VzJIB:-KOAA7VKxQhkTYXrwC6Z:-2063839539) (no related file):*
Very nice! 

Would be nice to be able to see a unified view of unit and integration tho. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/64)*
<!-- Sent from Reviewable.io -->

kwk commented:
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/almighty/almighty-core/64#-KOAA7VJ0kFmgj8VzJIB:-KOASHJfGkspu6ltOLyF:-700035763) (no related file):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> Very nice! 
> 
> Would be nice to be able to see a unified view of unit and integration tho. 
</details>
I'd be happy to do that later.

---

*[.gitignore, line 26 \[r1\]](https://reviewable.io:443/reviews/almighty/almighty-core/64#-KOA2iDuaG2i171tkEUA:-KOAPv0nw21G8FEcqrtR:1303854960) ([raw file](https://github.com/almighty/almighty-core/blob/2db23219938c373843c8e778e2458c80b73483ef/.gitignore#L26)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> Fresh already use tmp/ to store some state. Should we reuse that folder here as well?
</details>
Good point. I've fixed it. Now we have a `$(TMP_PATH)` that we can use inside Makefiles which points as an absolute path to the `tmp` in the source dir.

---

*[Makefile, line 55 \[r1\]](https://reviewable.io:443/reviews/almighty/almighty-core/64#-KOA40_h9FMhs3BAmeBS:-KOASN_7k63V_sS83V3A:-869122407) ([raw file](https://github.com/almighty/almighty-core/blob/9044a75dec88ae0ea999171c6206c7183137b5c3/Makefile#L55)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> Is it possible to define something like; clean-* ?
> 
> Would be nice if we could have dependencies in only one direction. Currently, if you comment out the test.mk the main mk fail due to missing targets.
</details>
Okay. I've introduced a global `$(CLEAN_TARGETS)` variable and whenever you add your own cleanup target to this by using `CLEAN_TARGETS += my-own-target`, then your `my-own-target` will be called when you call `make clean`.

---

*[.make/test.mk, line 30 \[r1\]](https://reviewable.io:443/reviews/almighty/almighty-core/64#-KOA39QY0GJrkZ8fR3Qy:-KOAVXW5ZqZw38E3DFdl:824057449) ([raw file](https://github.com/almighty/almighty-core/blob/9044a75dec88ae0ea999171c6206c7183137b5c3/.make/test.mk#L30)):*
Check the output of `go tool cover --help` to see why it is called `func` as opposed to `console`:

>  -func string
>    	output coverage profile information for each function
>  -html string
>    	generate HTML representation of coverage profile

I've updated the in-source documentation to reflect this.

Shall I still update to the shorter version for ease of use?

---

*[.make/test.mk, line 109 \[r1\]](https://reviewable.io:443/reviews/almighty/almighty-core/64#-KOA3apgCJuz-relcdNP:-KOAWGQkQYBxZRnZbaSv:353486831) ([raw file](https://github.com/almighty/almighty-core/blob/9044a75dec88ae0ea999171c6206c7183137b5c3/.make/test.mk#L109)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> These are fairly internal to the Makefile. Any way to 'exclude' the target from bash autocomplete?
</details>
None that I know of.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/64)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented:
*[.make/test.mk, line 30 \[r1\]](https://reviewable.io:443/reviews/almighty/almighty-core/64#-KOA39QY0GJrkZ8fR3Qy:-KOAYVerm8VXwjDmaJxp:-1531291977) ([raw file](https://github.com/almighty/almighty-core/blob/9044a75dec88ae0ea999171c6206c7183137b5c3/.make/test.mk#L30)):*
<details><summary><i>Previously, kwk (Konrad Kleine) wrote…</i></summary>
> Check the output of `go tool cover --help` to see why it is called `func` as opposed to `console`:
> 
> >  -func string
> >    	output coverage profile information for each function
> >  -html string
> >    	generate HTML representation of coverage profile
> 
> I've updated the in-source documentation to reflect this.
> 
> Shall I still update to the shorter version for ease of use?
> 
</details>
I would say yea. As func is currently the only console option anyway

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/64)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented:
*[.make/test.mk, line 109 \[r1\]](https://reviewable.io:443/reviews/almighty/almighty-core/64#-KOA3apgCJuz-relcdNP:-KOAZD5CuRObgpnyJWIp:754926811) ([raw file](https://github.com/almighty/almighty-core/blob/9044a75dec88ae0ea999171c6206c7183137b5c3/.make/test.mk#L109)):*
<details><summary><i>Previously, kwk (Konrad Kleine) wrote…</i></summary>
> None that I know of.
</details>
Prefixing the target with dot, same as how hidden files work, seems to work here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/64)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented:
*[a discussion](https://reviewable.io:443/reviews/almighty/almighty-core/64#-KOAA7VJ0kFmgj8VzJIB:-KOAZRtYiJecv7nsuTwK:1541278316) (no related file):*
<details><summary><i>Previously, kwk (Konrad Kleine) wrote…</i></summary>
> I'd be happy to do that later.
</details>
Sure, make a issue for it

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/64)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented:
Reviewed 4 of 7 files at r3.
Review status: 4 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/64)*
<!-- Sent from Reviewable.io -->

kwk commented:
Review status: 2 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[.make/test.mk, line 30 \[r1\]](https://reviewable.io:443/reviews/almighty/almighty-core/64#-KOA39QY0GJrkZ8fR3Qy:-KOdEQx9E58k5FRy1UOn:ba9hfaz) ([raw file](https://github.com/almighty/almighty-core/blob/9044a75dec88ae0ea999171c6206c7183137b5c3/.make/test.mk#L30)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> I would say yea. As func is currently the only console option anyway
</details>
Done with f4d5ebf96eaabb47810e5c21b527dc7c52252261

---

*[.make/test.mk, line 109 \[r1\]](https://reviewable.io:443/reviews/almighty/almighty-core/64#-KOA3apgCJuz-relcdNP:-KOdM6Sp4_J5Q-_rNRTC:bnqfnal) ([raw file](https://github.com/almighty/almighty-core/blob/9044a75dec88ae0ea999171c6206c7183137b5c3/.make/test.mk#L109)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> Prefixing the target with dot, same as how hidden files work, seems to work here.
</details>
Target is (no longer) listed with tab completion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/64)*
<!-- Sent from Reviewable.io -->
